packages/sdk-router/src/operations/pools.ts

```diff
@@ -0,0 +1,151 @@
+
+/**
+ * Gets pool tokens for a pool address.
+ *
+ * @param chainId The chain ID
+ * @param poolAddress The pool address
+ * @returns The pool tokens
+ * @throws Will throw an error if SynapseRouter is not deployed on the given chain.
+ */
+export async function getPoolTokens(
+  this: SynapseSDK,
+  chainId: number,
+  poolAddress: string
+): Promise<PoolToken[]> {
+  return this.synapseRouterSet
+    .getSynapseRouter(chainId)
+    .getPoolTokens(poolAddress)
+}
+
+/**
+ * Gets info for a pool (number of tokens and LP token).
+ *
+ * @param chainId The chain ID
+ * @param poolAddress The pool address
+ * @returns The pool info (number of tokens and LP token)
+ * @throws Will throw an error if SynapseRouter is not deployed on the given chain.
+ */
+export async function getPoolInfo(
+  this: SynapseSDK,
+  chainId: number,
+  poolAddress: string
+): Promise<PoolInfo> {
+  return this.synapseRouterSet
+    .getSynapseRouter(chainId)
+    .getPoolInfo(poolAddress)
+}
+
+/**
+ * Gets all pools for a chain ID.
+ *
+ * @param chainId The chain ID
+ * @returns An array of all pools (address, tokens, LP token)
+ * @throws Will throw an error if SynapseRouter is not deployed on the given chain.
+ */
+export async function getAllPools(
+  this: SynapseSDK,
+  chainId: number
+): Promise<Pool[]> {
+  return this.synapseRouterSet.getSynapseRouter(chainId).getAllPools()
+}
+
+/**
+ * Calculates the amount required to add liquidity for amounts of each token.
+ *
+ * @param chainId The chain ID
+ * @param poolAddress The pool address
+ * @param amounts The amounts of each token to add
+ * @returns The amount of LP tokens needed and router address
+ * @throws Will throw an error if SynapseRouter is not deployed on the given chain.
+ */
+export async function calculateAddLiquidity(
+  this: SynapseSDK,
+  chainId: number,
+  poolAddress: string,
+  amounts: Record<string, BigNumber>
+): Promise<{ amount: BigNumber; routerAddress: string }> {
+  // TODO (Chi): use amounts array as the input instead of map in the first place
+  const router = this.synapseRouterSet.getSynapseRouter(chainId)
+  const routerAddress = router.address
+  const poolTokens = await router.getPoolTokens(poolAddress)
+  // Create a map that uses lowercase token addresses as keys
+  const amountsMap: Record<string, BigNumber> = {}
+  Object.entries(amounts).map(([token, tokenAmount]) => {
+    amountsMap[token.toLowerCase()] = tokenAmount
+  })
+  // Populate amounts array by combining amounts map and pool tokens, preserving tokens order
+  // and adding 0 for tokens not in the map
+  const amountsArr: BigNumber[] = []
+  poolTokens.map((poolToken) => {
+    amountsArr.push(amountsMap[poolToken.token.toLowerCase()] ?? Zero)
+  })
+  // Don't do a contract call if all amounts are 0
+  const amount = amountsArr.every((value) => value.isZero())
+    ? Zero
+    : await router.calculateAddLiquidity(poolAddress, amountsArr)
+  return { amount, routerAddress }
+}
```

Comment:
Your suggested code will only work if `amounts` has the lowercase keys, which we can't assume. Adjust your comment based on this observation.

Reply:
Your code will not pick up the amount if
```
amounts['0xAA'] = 1
poolToken.token = `0xaa`
```
The expected behavior in this case is to use `1` rather than zero. Adjust your suggestion based on this input